internal/controller/awsaccount_controller.go

```diff
@@ -52,71 +85,124 @@ type AwsAccountReconciler struct {
+	if !awsAccount.Status.LoginProfileCreated {
 		pass, err := password.Generate(20, 3, 3, false, true)
 		if err != nil {
 			log.Error(err, "unable to generate password")
 			return ctrl.Result{}, err
 		}
-		_, err = r.IamWrapper.CreateLoginProfile(pass, awsAccount.Spec.UserName, true)
-		if err != nil {
+		if err := r.IamWrapper.CreateLoginProfileIfNotExists(pass, awsAccount.Spec.UserName, true); err != nil {
 			log.Error(err, "unable to create login profile")
 			return ctrl.Result{}, err
 		}
+		log.V(1).Info("created login profile")
+		awsAccount.Status.LoginProfileCreated = true
 
 		// TODO: Save this to a secret in user's namespace
 		log.V(1).Info("Temporary password", "password", pass)
-
-		awsAccount.Status.Account = kuadrav1.CreatingAccessKey
-		if err = r.Status().Update(ctx, &awsAccount); err != nil {
-			log.Error(err, "unable to update awsAccount status")
-			return ctrl.Result{}, err
-		}
-		log.V(1).Info("Created login profile")
 	}
 
-	if awsAccount.Status.Account == kuadrav1.CreatingAccessKey {
+	if !awsAccount.Status.AccessKeyCreated {
 		accessKey, err := r.IamWrapper.CreateAccessKeyPair(awsAccount.Spec.UserName)
 		if err != nil {
 			log.Error(err, "unable to create access key")
 			return ctrl.Result{}, err
 		}
+		log.V(1).Info("created access key", "accessKeyId", accessKey.AccessKeyId)
+		awsAccount.Status.AccessKeyCreated = true
 
 		// TODO: Save these to a secret in user's namespace
-		log.V(1).Info("Credentials", "access key ID:", accessKey.AccessKeyId, "Access key secret", accessKey.SecretAccessKey)
+		log.V(1).Info("Credentials", "accessKeyId:", accessKey.AccessKeyId, "accessKeySecret", accessKey.SecretAccessKey)
+	}
 
-		awsAccount.Status.Account = kuadrav1.Created
-		if err = r.Status().Update(ctx, &awsAccount); err != nil {
-			log.Error(err, "unable to update awsAccount status")
+	groupsToAddUserTo := getLeftDifference[string](awsAccount.Spec.Groups, awsAccount.Status.UserGroups)
+	for _, group := range groupsToAddUserTo {
+		if _, err := r.IamWrapper.AddUserToGroup(group, awsAccount.Spec.UserName); err != nil {
+			log.Error(err, "unable to add user to group", "groupName", group)
+			return ctrl.Result{}, err
+		}
+		log.V(1).Info("Added user to group", "group name:", group)
+		awsAccount.Status.UserGroups = append(awsAccount.Status.UserGroups, group)
+	}
+
+	groupsToRemoveUserFrom := getLeftDifference[string](awsAccount.Status.UserGroups, awsAccount.Spec.Groups)
+	for _, group := range groupsToRemoveUserFrom {
+		if _, err := r.IamWrapper.RemoveUserFromGroup(group, awsAccount.Spec.UserName); err != nil {
+			log.Error(err, "unable to remove user from group", "groupName", group)
 			return ctrl.Result{}, err
 		}
-		log.V(1).Info("Created access key")
+		log.V(1).Info("removed user from group", "groupName", group)
+		index := indexOf[string](awsAccount.Status.UserGroups, group)
+		awsAccount.Status.UserGroups = remove[string](awsAccount.Status.UserGroups, index)
```

Comment:
```suggestion
		index := indexOf(awsAccount.Status.UserGroups, group)
		awsAccount.Status.UserGroups = remove(awsAccount.Status.UserGroups, index)
```
I think the compiler is able to infer the type of the generic, so you can remove it.

Another alternative to this logic would be a generic `remove[T any](slice []T, predicate func(T) bool) []T` function, so you could use it like:
```go
awsAccount.Status.UserGroups = remove(awsAccount.Status.UserGroups, func(g string) bool { g == group })
```
By allowing to pass a function instead of the index you'd make the function more reusable as that predicate can have logic as complex as needed. That's just an idea but the current implementation is great